lib/allocator.hpp

```diff
@@ -0,0 +1,215 @@
+#ifndef __ALLOCATOR_INCLUDED
+#define __ALLOCATOR_INCLUDED
+
+#include <cstdint>
+#include <cstdlib>
+
+#include <new>
+#include <stdexcept>
+#include <memory>
+#include <utility>
+#include <limits>
+#include <type_traits>
+
+namespace usr {
+
+/* Allocator stuff
+ */
+
+namespace internal {
+
+template <typename X, std::size_t N>
+class SpecialAllocator {
+private:
+	using idx_t =
+		typename std::conditional<N <= std::numeric_limits<uint8_t>::max(),
+			uint8_t,
+			typename std::conditional<N <= std::numeric_limits<uint16_t>::max(),
+				uint16_t,
+				typename std::conditional<N <= std::numeric_limits<uint32_t>::max(),
+					uint32_t,
+					std::size_t
+				>::type
+			>::type
+		>::type;
+
+	static_assert(sizeof(X) >= sizeof(idx_t),
+		"N is too large for type X: sizeof(<the minimal underlying unsigned integral type for N>) must be not larger than sizeof(X)"
+	);
+
+	class Data {
+	private:
+/* Shame on Microsoft: the only one who does not support untagged unions */
+#ifdef _MSC_VER
+		union U {
+#else
+		union {
+#endif
+			X val;
+			idx_t idx;
+		} (*buf)[N];
+
+		idx_t current;
+		idx_t filled;
+
+	public:
+		Data() noexcept: buf{}, current{}, filled{} {
+		}
+
+		~Data() {
+			if (buf) {
+				std::free(buf);
+			}
+		}
+
+		X *allocate() {
+			if (!buf) {
+				if (!(buf = reinterpret_cast<decltype(buf)>(std::malloc(sizeof *buf)))) {
+					throw std::bad_alloc();
+				}
```

Comment:
1. Что значит, "каждый раз проверять возвращаемое значение"?
Один раз написал и "забыл". В коде вообще один раз встречается.

"Каждый раз проверять возвращаемое значение" будет процессор, когда будет исполнять этот код.

В задании упомянуто, что с помощью `ltrace` предполагается контролировать операции выделения/освобождения памяти, и я сознательно выбрал `malloc`/`free`, чтобы это было сделать проще, поскольку я и сам должен был при отладке контролировать эти операции.
У утилиты `ltrace` есть опция `-e`, которая указывает утилите отфильтровать нужные функции, не показывая остальные. Для `operator'ов new/elete`, чтобы воспользоваться данной функциональностью, пришлось бы искать, как выглядят их mangled-имена.

И в чём состоит "большая правильность", что это такое?

2. Что значит "достаточно сильное" преобразование типов?
Почему, даже если это так, отсюда следует, что его не следует здесь применять?

[Здесь](https://en.cppreference.com/w/cpp/language/reinterpret_cast) в самом начале написано, для чего используют `reinterpret_cast`:
>Unlike static_cast, but like const_cast, the reinterpret_cast expression does not compile to any CPU instructions (except when converting between integers and pointers or on obscure architectures where pointer representation depends on its type).

И далее:

> It is purely a compile-time directive which instructs the compiler to treat expression as if it had the type new_type.

Именно то, что мне здесь и нужно: никаких преобразований и "настроек" указателя.

> В вашем случае идет преобразование указателя void* в другой указатель. И хотя другой указатель ссылается на сложный тип, с таким преобразованием прекрасно справляется static_cast его в этом случае достаточно.

Мне нужно вернуть численно то же самое значение адреса, которое вернула `malloc`. Это одна из тех вещей, для которых, в частности, и был создан `reinterpret_cast`. Поэтому именно его я здесь и использую.

Reply:
1) На счет ltrace  вы правы - это усложняет задачу анализа. И все же не сильно, в одном случае вы выделяете 10 раз скажем по 16  байт, а в другом 160 байт за раз. Такой паттерн все же не так сложно распознать. 
В самом задании не говорится о том, каким способом вы будите выделять память. Так что можно malloc и не использовать.
Что касается "большей правильности", то под этим я имел виду:
 а.  Это все же средство С++,  а не С 
 b.  Вам не нужно проверять возвращаемое значение
 с. Не везде вы сможете его элегантно проверить, не применительно к нашему случаю, но если мы выделяем память в списке инициализации, проверять там возвращаемое значение - то еще удовольствие.
"Один раз написал и "забыл". В коде вообще один раз встречается."
"и я сознательно выбрал malloc/free"
Ok. Если вы выбрали это осознано - это меняет дело. Но не отметить это я не мог, сами понимаете.

2) Что значит "достаточно сильное" преобразование типов?
Почему, даже если это так, отсюда следует, что его не следует здесь применять?

Потому же почему не следует применять преобразования в стиле С. Которые еще более сильны, чем reinterpret_cast, и более лаконичны. И с другой стороны, подумайте, зачем вообще в языке существует static_cast, если по вашей логике всегда можно воспользоваться более сильным преобразованием? 

Очевидно, чтобы не допустить ошибки, всегда следует выбирать более слабое преобразование. К примеру, после рефакторинга reinterpret_cast не заметит ошибки, static_cast же может и не скомпилировать некоторые случаи, что лишний раз заставит подумать, правильно ли мы что-то делаем.
Я просто оставлю это здесь, там большое количество аргументов в защиту моей позиции:
https://stackoverflow.com/questions/310451/should-i-use-static-cast-or-reinterpret-cast-when-casting-a-void-to-whatever

https://stackoverflow.com/questions/332030/when-should-static-cast-dynamic-cast-const-cast-and-reinterpret-cast-be-used

https://stackoverflow.com/questions/573294/when-to-use-reinterpret-cast
А вот любопытная цитата одного товарища: " In principle, it is possible to use the result of a static_cast without casting it back to its original type, whereas you should always cast the result of a reinterpret_cast back to its original type before using it to ensure portability."Bjarne Stroustrup’s  Взял отсюда https://isocpp.org/wiki/faq/style-and-techniques#static-cast

А вот еще тот же автор детально раскрывает свою мысль [ссылка ](https://books.google.ru/books?id=hS9mDwAAQBAJ&pg=PT390&lpg=PT390&dq=pointers+reinterpret_cast+vs+static_cast+Stroustrup&source=bl&ots=xNld7p8-Sm&sig=ACfU3U3VIkl6hNnKYqBL18fzlRH1fry8Xg&hl=ru&sa=X&ved=2ahUKEwig36C-rI3hAhVIAxAIHch_Bn84ChDoATADegQIBxAB#v=onepage&q=pointers%20reinterpret_cast%20vs%20static_cast%20Stroustrup&f=false) не знаю откроется ли ссылка в нужном месте

А вот еще цитата из Bjarne Stroustrup’s proposal  от 1993 года "In addition, static_cast can cast from any pointer to object type to a void* and from void* to any pointer to object type. The void* to T* conversion cannot be checked but is well-defined." в котором он говорит, что для чего нужно [ссылка](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/1993/N0349a.pdf)

Reply:
1. В дополнение -- `::operator new[]` можно перегрузить, и аллокатор начнёт "брать память" из другого места и другим способом. Насколько это правильно, это ещё "тот" вопрос.
2. Вы ссылаетесь на авторитетов, которых для меня не существует. Что-то может быть так-то потому-то и потому-то, а не потому что так сказал кто-то авторитетный. Он может ошибиться. Также так и не понятна "сила" или "слабость" приведений. Есть "преобразующие", `static_cast`, который при преобразовании из указателя на один класс в указатель на другой класс, особенно, если они связаны наследованием, и оно множественное, может изменить численное значение адреса, выполнив его "настройку", есть меняющие точку зрения на область памяти, reinterpret_cast, и никаких "настроек" и прочих "вычислений" не производится, `const_cast`, для смены квалификаторов типов, и C-style cast, который представляет из себя целый агрегат попыток применения различных видов cast'ов по определённому алгоритму. Никаких "сил" или "слабостей" нет, есть определённые свойства у каждой операции.

[Видите](https://wandbox.org/permlink/Wiagy4cAdqLW3UWt), чем отличается поведение `reinterpret_cast` от поведения `static_cast`? Вот, мне нужно поведение именно `reinterpret_cast`. Смена интерпретации области памяти при неизменном адресе. А `static_cast` может ещё и временный объект создать, как в данном примере. Опять же, нет никакой "силы" или "слабости", есть вполне определённые свойства.

Более того, память ещё не инициализирована, её даже читать нельзя, предварительно не записав, а `static_cast` об этом не знает. То есть, это указатель на не совсем настоящий тип `X` или, даже, на совсем не настоящий тип `X`, о чём, кстати, говорилось на лекции.

Reply:
1. Есть много способов выстрелить себе в ногу.
::operator new[] можно перегрузить, а можно ведь и не перегружать. Ведь так? Это фича, и полезная, но это не значит, что вы должны ее использовать.

2. В мире есть много чего. Не горю желанием обсуждать все, что есть в мире. У вас есть приведение void* в указатель на другой тип. Для этого рекомендуется использовать static_cast. Рекомендация выработана не мной, а сообществом. Ссылки на рекомендации я предоставил, объяснение почему так лучше дано там же.

“Вы ссылаетесь на авторитетов, которых для меня не существует. “ – это не доказательство вашей правоты или моей не правоты, это всего лишь манифест вашего отношение к жизни. За кружкой пива я готов порассуждать об этом, но не здесь.

В данном случае, если вы считаете, что ваша позиция более верная, чем аргументация “авторитетов”, единственное, что имеет смысл, это изложить ее, а не громкие заявления.

“Вот, мне нужно поведение именно reinterpret_cast.” Вам не нужно это поведение в вашем случае. Вы хотите: “Смена интерпретации области памяти при неизменном адресе.” С помощью reinterpret_cast вы выстрелите себе в ногу, поскольку C++ не гарантирует вам этого [std C++17 8.2.10/3]: 

“The mapping performed by reinterpret_cast might, or might not, produce a representation
 different from the original value.” 

Напротив, в случае приведение void* в указатель на другой тип стандарт не только  говорит, что это преобразование осуществляет static_cast, но и гарантирует неизменность значения адреса у полученного указателя [std C++17 8.2.9/13]:

“**A prvalue of type “pointer to cv1 void” can be converted to a prvalue of type “pointer to cv2 T”,
 where T is an object type and cv2 is the same cv-qualification as, or greater cv-qualification than,
 cv1.** If the original pointer value represents the address A of a byte in memory and A does not
 satisfy the alignment requirement of T, then the resulting pointer value is unspecified. Otherwise,
 if the original pointer value points to an object a, and there is an object b of type T (ignoring cv-
qualification) that is pointer-interconvertible (6.9.2) with a, the result is a pointer to b. **Otherwise,
 the pointer value is unchanged by the conversion.**” 

На всякий случай, согласно [std C++17 6.7.4.1/2] аллоцирующие ф-ии (operator new и operator new[]) не требуют беспокойства по поводу выравнивания памяти:

“The pointer returned shall be suitably aligned so that it can be converted to a pointer to any
suitable complete object type (21.6.2.1) and then used to access the object or array in the storage
allocated(until the storage is explicitly deallocated by a call to a corresponding deallocation
function).”

Пример который вы рассматриваете, не корректен. Мы обсуждаем приведение указателей. К нашему случаю это не имеет никакого отношения. Для вашего случая приведение к ссылке есть специальное правило, которое создает временный объект [std C++17 8.2.9/4]. Других случаев создания временных объектов я не припомню.

Я привел вам положения стандарта, очень надеюсь, что хоть к стандарту вы относитесь иначе, чем к “авторитетам”. Рекомендую иногда его открывать.

Reply:
1. Я о том, что в случае использования `::operator new[]`, поведение аллокатора начинает зависеть от того, переопределён он или нет. А в случае использования `malloc` такой зависимости нет.
2. Опять сообщество.

Я должен как-то дать вам понять, что ссылки на авторитеты не являются для меня аргументами. Свою позицию по поводу `reinterpret_cast`, почему я использую именно его, я изложил несколько раз.

>С помощью reinterpret_cast вы выстрелите себе в ногу, поскольку C++ не гарантирует вам этого [std C++17 8.2.10/3]:
>
>“The mapping performed by reinterpret_cast might, or might not, produce a representation
different from the original value.”

Вы вырвали цитату из контекста:

>2 The reinterpret_cast operator shall not cast away constness (8.2.11). An expression of integral, enumeration, pointer, or pointer-to-member type can be explicitly converted to its own type; such a cast yields the value of its operand.
3 [ Note: The mapping performed by reinterpret_cast might, or might not, produce a representation different from the original value. — end note ]

Во-первых, это относится к преобразованию в тот же самый тип, что и исходный.
Во-вторых, если одно и то же значение может иметь различные представления, то возможно, мне придётся это учитывать (а, возможно, и нет), -- только и всего.

Но у меня другой случай, я преобразую в другой тип.

Если же прочитать далее [std C++17 8.2.10/7], то можно выяснить, что `reinterpret_cast` я применяю правомерно:

>7 An object pointer can be explicitly converted to an object pointer of a different type. [73] When a prvalue v of object pointer type is converted to the object pointer type “pointer to cv T”, the result is static_cast<cv T*>(static_cast<cv void*>(v)). [ Note: Converting a prvalue of type “pointer to T1” to the type “pointer to T2” (where T1 and T2 are object types and where the alignment requirements of T2 are no stricter than those of T1) and back to its original type yields the original pointer value. — end note ]

Этот пункт явно описывает мой случай. Если "раскрыть" этот пункт, у меня `void *` сначала приводится к `void *`, а потом к `decltype(buf)`.

>Напротив, в случае приведение void* в указатель на другой тип стандарт не только говорит, что это преобразование осуществляет static_cast, но и гарантирует неизменность значения адреса у полученного указателя [std C++17 8.2.9/13]

Хорошо, но разве этот факт запрещает в данном месте применить `reinterpret_cast`?

И вы оборвали этот пункт, там же дальше пример идёт, в котором, фактически, записана строка из [std C++17 8.2.10/7]:

>13 ... Otherwise, the pointer value is unchanged by the conversion. [ Example:
T* p1 = new T;
const T* p2 = static_cast<const T*>(static_cast<void*>(p1));
bool b = p1 == p2; // b will have the value true.
— end example ]

Повторю вопрос: есть что-то, что запрещает мне применить здесь `reinterpret_cast`?

>Пример который вы рассматриваете, не корректен.

Он корректен в том смысле, что показывает, насколько `static_cast` может быть "вычисляющим" преобразованием.

>Мы обсуждаем приведение указателей.

[Пример](https://wandbox.org/permlink/uDCZQMjjHn2vz05d) с "настройкой" указателей.

>Я привел вам положения стандарта, очень надеюсь, что хоть к стандарту вы относитесь иначе, чем к “авторитетам”. Рекомендую иногда его открывать.

Стандарт является первоисточником и "последней инстанцией", правомерные ссылки на стандарт являются самыми сильными аргументами.

Reply:
Эээ. Т.е. есть положения стандарта о  static_cast,, где есть конкретный пункт преобразование void* в указатель на T - вы его игнорируете. Есть пункт который говорит, что  reinterpret_cast не дает гарантий после преобразования об идентичность представления значений (вынесенный в отдельный пункт!).  Есть создатель языка и автор этих положений стандарта - Страуструп, который говорит что: "чуваки не делайте так, потому что вот есть стандарт, где я написал  - что написал." 
И тут приходите Вы, и говорите: "да что там Страуструп, подумаешь авторитет. Я лучше знаю, как интерпретировать стандарт" ?
Я даже не хочу гадать, что с вами не так. Буду исходить из того, что вы меня жестко троллите. Простите, но время время... Я с вашего позволения займусь чем-нибудь еще. Дальнейшее обсуждение, конкретно этой работы, считаю со своей стороны не продуктивным.

Reply:
>Эээ. Т.е. есть положения стандарта о static_cast,, где есть конкретный пункт преобразование void* в указатель на T - вы его игнорируете.

Вы хотите сказать, что стандарт от меня **требует** преобразовать имено так и никак иначе? Иначе говоря, **запрещает** по-иному пробразовывать?

Повторю вопрос в третий и последний раз: есть ли что-то в стандарте, что запрещает мне применить здесь `reinterpret_cast`?

>Есть пункт который говорит, что reinterpret_cast не дает гарантий после преобразования об идентичность представления значений (вынесенный в отдельный пункт!).

Вы читали мой ответ по этому поводу?
К какому приведению это относится? Мой ли это случай?
Даже если бы был мой, -- видели вопрос по поводу того, чем мне грозит неидентичность представления?
Можете привести пример неидентичности представления?

Вы читали [std C++17 8.2.10/7]?
Могу ли я выполнить такое приведение?
```cpp
buf = static_cast<decltype(buf)>(static_cast<void*>(std::malloc(sizeof *buf)));
```
Если могу, и вы читали [std C++17 8.2.10/7], то понимаете ли вы, что последнее приведение эквивалентно вот этому приведению?
```cpp
buf = reinterpret_cast<decltype(buf)>(std::malloc(sizeof *buf));
```

>Есть создатель языка и автор этих положений стандарта - Страуструп, который говорит что: "чуваки не делайте так, потому что вот есть стандарт, где я написал - что написал."

Стандарт не Страуструп единолично пишет. Страуструп есть в жизни, но когда речь заходит о C++, есть только стандарт. Кто там что говорит, вообще к делу не относится.

>И тут приходите Вы, и говорите: "да что там Страуструп, подумаешь авторитет. Я лучше знаю, как интерпретировать стандарт" ?

Стандарт не должен быть таким, что только Страуструп его может правильно понять, а то это превращается в какую-то секту свидетелей Страуструпа. Вы считаете, что стандарт нельзя читать самому, а вместо этого следует читать интерпретации стандарта, изложенные Страуструпом?

>Дальнейшее обсуждение, конкретно этой работы, считаю со своей стороны не продуктивным.

Вы же не видите моих аргументов. Я вынужден по три раза вопрос задавать. 
Вы так и не ответили на вопрос, почему я не могу применить `reinterpret_cast`, запрещает ли мне стандарт его использование.

Ответ, что, мол, Страуструп сказал всем использовать здесь `static_cast`, ... я даже слов не могу подобрать. Нет для C++-программ никакого Страуструпа, есть только стандарт.

Действительно, не продуктивно.